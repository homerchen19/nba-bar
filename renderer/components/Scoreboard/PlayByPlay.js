import React from 'react';
import PropTypes from 'prop-types';
import { Flex } from 'antd-mobile';
import styled from 'styled-components';
import { AutoSizer, MultiGrid } from 'react-virtualized';

import { colors } from '../../styles/theme';
import nba from '../../utils/nba';

const Wrapper = styled(Flex)`
  width: 100%;
  height: 340px;
  background: #fff;
`;

const StyledCell = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin: 0;
  padding: 0;
  overflow: auto;
  border-bottom: 1px solid ${colors.white};
  text-align: ${props => props.align};
  word-wrap: break-word;
  white-space: pre;
`;

const Score = styled.span`
  color: ${props => props.color};
  font-weight: ${props => (props.color === '#000' ? 200 : 700)};
`;

const getQuaterString = period =>
  period > 5 ? `OT${period - 4}` : `Q${period}`;

const getClock = clock => (clock !== '' ? clock : '12:00');

const PlayByPlay = ({ gamePlayByPlayData }) => (
  <Wrapper justify="center">
    <Flex.Item style={{ height: '100%' }}>
      <AutoSizer>
        {({ height }) => (
          <MultiGrid
            fixedColumnCount={1}
            estimatedColumnSize={300}
            columnWidth={({ index }) => (index === 0 ? 100 : 200)}
            columnCount={2}
            cellRenderer={({ columnIndex, key, rowIndex, style }) => {
              const data = gamePlayByPlayData[rowIndex];

              let homeColor = '#000';
              let visitorColor = '#000';

              if (rowIndex !== gamePlayByPlayData.length) {
                const previousData = gamePlayByPlayData[rowIndex + 1];
                if (data.home_score > previousData.home_score) {
                  homeColor = nba.getTeamBackgroundColor(data.team_abr);
                } else if (data.visitor_score > previousData.visitor_score) {
                  visitorColor = nba.getTeamBackgroundColor(data.team_abr);
                }
              }

              if (columnIndex === 0) {
                return (
                  <StyledCell key={key} align="center" style={style}>
                    <p>
                      {`${getQuaterString(+data.period)} ${getClock(
                        data.clock
                      )}`}
                    </p>
                    <p>
                      <Score key="homeScore" color={homeColor}>
                        {data.home_score}
                      </Score>{' '}
                      -{' '}
                      <Score key="visistorScore" color={visitorColor}>
                        {data.visitor_score}
                      </Score>
                    </p>
                  </StyledCell>
                );
              }

              return (
                <StyledCell key={key} align="left" style={style}>
                  {data.description.replace(/\[.*\]/i, '').trim()}
                </StyledCell>
              );
            }}
            rowCount={gamePlayByPlayData.length}
            rowHeight={50}
            width={300}
            height={height}
            hideTopRightGridScrollbar
            hideBottomLeftGridScrollbar
          />
        )}
      </AutoSizer>
    </Flex.Item>
  </Wrapper>
);

PlayByPlay.propTypes = {
  gamePlayByPlayData: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default PlayByPlay;
