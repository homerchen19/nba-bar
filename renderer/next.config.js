const withCSS = require('@zeit/next-css');
const withImages = require('next-images');

module.exports = withImages(
  withCSS({
    cssModules: true,
    webpack(config, { isServer }) {
      config.target = 'electron-renderer';

      if (isServer) {
        const antStyles = /antd-mobile\/.*?\/style.*?/;
        const origExternals = [...config.externals];
        config.externals = [
          (context, request, callback) => {
            if (request.match(antStyles)) return callback();
            if (typeof origExternals[0] === 'function') {
              origExternals[0](context, request, callback);
            } else {
              callback();
            }
          },
          ...(typeof origExternals[0] === 'function' ? [] : origExternals),
        ];

        config.module.rules.unshift({
          test: antStyles,
          use: 'null-loader',
        });
      }

      return config;
    },
    exportPathMap() {
      return {
        '/home': { page: '/home' },
        '/live': { page: '/live' },
        '/preview': { page: '/preview' },
        '/scoreboard': { page: '/scoreboard' },
        '/settings': { page: '/settings' },
        '/standings': { page: '/standings' },
      };
    },
  })
);
