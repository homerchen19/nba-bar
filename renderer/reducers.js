import { combineReducers } from 'redux';

import homeReducer from './containers/Home/reducer';
import previewReducer from './containers/Preview/reducer';
import liveReducer from './containers/Live/reducer';
import scoreboardReducer from './containers/Scoreboard/reducer';
import standingsReducer from './containers/Standings/reducer';
import settingsReducer from './containers/Settings/reducer';

const rootReducer = combineReducers({
  home: homeReducer,
  preview: previewReducer,
  live: liveReducer,
  scoreboard: scoreboardReducer,
  standings: standingsReducer,
  settings: settingsReducer,
});

export default rootReducer;
