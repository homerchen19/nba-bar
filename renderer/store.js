import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';

import rootReducer from './reducers';

const middleware = [thunk, logger];

const enhancers = compose(
  applyMiddleware(...middleware),
  typeof window !== 'undefined' &&
    window.devToolsExtension &&
    process.env.NODE_ENV !== 'production'
    ? window.devToolsExtension()
    : f => f
);

export default function configureStore(initialState = {}) {
  const store = createStore(rootReducer, initialState, enhancers);

  return store;
}
