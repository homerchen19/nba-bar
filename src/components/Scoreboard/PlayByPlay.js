import React from 'react';
import PropTypes from 'prop-types';
import R from 'ramda';
import { Flex } from 'antd-mobile';
import styled from 'styled-components';
import { StickyTable, Row, Cell } from 'react-sticky-table';

import { colors } from '../../styles/theme';
import nba from '../../utils/nba';

const Wrapper = styled(Flex)`
  width: 100%;
  height: 340px;
  background: #fff;
`;

const StyledCell = styled(Cell)`
  padding: 8px 11px;
  border-bottom: 1px solid ${colors.white};
  text-align: ${props => props.align};
  word-wrap: break-word;
  white-space: pre;
`;

const Score = styled.span`
  color: ${props => props.color};
  font-weight: ${props => (props.color === '#000' ? '200' : '700')};
`;

const getQuaterString = period =>
  period > 5 ? `OT${period - 4}` : `Q${period}`;

const getClock = clock => (clock !== '' ? clock : '12:00');

const latestScore = {
  home: 0,
  visitor: 0,
};

const reanderPlayByPlayRow = R.compose(
  R.reverse,
  R.map(data => {
    let homeColor = '#000';
    let visitorColor = '#000';

    if (+data.home_score > latestScore.home) {
      homeColor = nba.getTeamBackgroundColor(data.team_abr);
    } else if (+data.visitor_score > latestScore.visitor) {
      visitorColor = nba.getTeamBackgroundColor(data.team_abr);
    }

    latestScore.home = +data.home_score;
    latestScore.visitor = +data.visitor_score;

    return (
      <Row key={data.event}>
        <StyledCell key="name" align="center">
          <p>{`${getQuaterString(+data.period)} ${getClock(data.clock)}`}</p>
          <p>
            <Score color={homeColor}>{data.home_score}</Score> -{' '}
            <Score color={visitorColor}>{data.visitor_score}</Score>
          </p>
        </StyledCell>
        <StyledCell key="score" align="left" style={{ padding: '8px 0' }}>
          {data.description.replace(/\[.*\]/i, '').trim()}
        </StyledCell>
      </Row>
    );
  })
);

const PlayByPlay = ({ gamePlayByPlayData }) => (
  <Wrapper justify="center">
    <StickyTable stickyHeaderCount={0}>
      {reanderPlayByPlayRow(gamePlayByPlayData)}
    </StickyTable>
  </Wrapper>
);

PlayByPlay.propTypes = {
  gamePlayByPlayData: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default PlayByPlay;
