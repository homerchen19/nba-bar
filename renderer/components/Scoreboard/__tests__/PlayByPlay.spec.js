import React from 'react';
import renderer from 'react-test-renderer';

import PlayByPlay from '../PlayByPlay';

const realMath = Object.create(global.Math);

beforeAll(() => {
  const mockMath = Object.create(global.Math);
  mockMath.random = () => 0.5;
  global.Math = mockMath;
});

afterAll(() => {
  global.Math = realMath;
});

it('be defined', () => {
  expect(PlayByPlay).toBeDefined();
});

it('match snapshot', () => {
  const props = {
    gamePlayByPlayData: [
      {
        event: '0',
        clock: '',
        description: 'Start Period',
        player_code: '',
        person_id: '',
        home_score: '0',
        visitor_score: '0',
        team_abr: '',
        period: '1',
      },
      {
        event: '1',
        clock: '',
        description: 'Jump Ball Adams vs Duncan (Green gains possession)',
        player_code: 'steven_adams',
        person_id: '203500',
        home_score: '0',
        visitor_score: '0',
        team_abr: 'OKC',
        period: '1',
      },
      {
        event: '2',
        clock: '11:43',
        description: '[OKC] Adams Foul: Shooting (1 PF) (2 FTA) (B Spooner)',
        player_code: 'steven_adams',
        person_id: '203500',
        home_score: '0',
        visitor_score: '0',
        team_abr: 'OKC',
        period: '1',
      },
    ],
  };
  const tree = renderer.create(<PlayByPlay {...props} />).toJSON();

  expect(tree).toMatchSnapshot();
});
