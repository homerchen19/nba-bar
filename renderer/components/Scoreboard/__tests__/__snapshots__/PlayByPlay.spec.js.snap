// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
<div
  className="am-flexbox PlayByPlay__Wrapper-sc-1eahhxw-0 iuAhTg am-flexbox-justify-center am-flexbox-align-center"
>
  <div
    className="am-flexbox-item"
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    />
  </div>
</div>
`;
